File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = lambda: sys.stdin.readline().rstrip()

def check(s: str):
    stk = []
    for c in s:
        if c == ')':
            if not stk or stk[-1] == '[':
                return False
            else:
                stk.pop()

        if c == ']':
            if not stk or stk[-1] == '(':
                return False
            else:
                stk.pop()

        if c == '(' or c == '[':
            stk.append(c)

    if stk: return False
    else: return True

while True:
    s = input()
    if s == ".":
        break

    ret = check(s)
    if ret:
        print("yes")
    else:
        print("no")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    while (true) {
        string s;
        getline(cin, s);
        if (s == ".") break;

        stack<char> stk;
        bool check = true;
        for (int i=0; i<s.length(); i++) {
            if (s[i] == ')') {
                if (stk.size() == 0 || stk.top() == '[') {
                    check = false;
                    break;
                } else {
                    stk.pop();
                }
            }

            if (s[i] == ']') {
                if (stk.size() == 0 || stk.top() == '(') {
                    check = false;
                    break;
                } else {
                    stk.pop();
                }
            }

            if (s[i] == '(') stk.push(s[i]);
            if (s[i] == '[') stk.push(s[i]);
        }
        if (stk.empty() && check) cout << "yes\n";
        else cout << "no\n";
    }
    return 0;
}